### ISSUE TITLE
Unparsed atoms with json parsing element sometimes end in the middle of elements

### ISSUE THREAD

**Issue by landauermax** [ait-aecid/logdata-anomaly-miner#603]
I noticed that sometimes, the information on where the parsing failed is strange. Example:

```
2021-04-02 10:11:33 Unparsed atom received
VerboseUnparsedAtomHandler: "None" (1 lines)
  {
  "DebugLog": [
    "Starting match update on b'{\"a\": \"fails\", \"x\": {\"z\": \"test\", \"y\": \"string fails because of colon: end of string\", \"b\": null}}'",
    "Removed b'{\"a\": \"fails', remaining 86 bytes",
    "Removed b'\", \"x\": {\"z\": \"test', remaining 67 bytes",
    "Removed b'\", \"y\": \"string fails because of colon: end', remaining 24 bytes",
    "Shortest unmatched data was b' of string\", \"b\": null}}'",
    ""
  ],
  "LogData": {
    "RawLogData": [
      "{\"a\": \"fails\", \"x\": {\"z\": \"test\", \"y\": \"string fails because of colon: end of string\", \"b\": null}}"
    ],
    "Timestamps": [
      1617358293.01
    ],
    "DetectionTimestamp": 1617358293.01,
    "LogLinesCount": 1
  },
  "AnalysisComponent": {
    "AnalysisComponentIdentifier": null,
    "AnalysisComponentType": "VerboseUnparsedAtomHandler",
    "AnalysisComponentName": null,
    "Message": "Unparsed atom received",
    "PersistenceFileName": null
  }
}
```

The "shortest unmatched data" starts with "of string", incorrectly indicating that "string fails because of colon: end" that was parsed by the same VariableByteDataModelElement was correctly parsed. This is confusing, especially in more complicated json logs it is difficult to tell which element is actually responsible for the unparsed event, and should be fixed. E.g., I would expect that the "shortest unmatched data" is either \"string fails because of colon: end of string\", \"b\": null}}" or even \"y\": \"string fails because of colon: end of string\", \"b\": null}}"

To reproduce the anomaly from above, use the following config and data:

```
LearnMode: True # optional
AminerUser: 'aminer'  # optional default: aminer
AminerGroup: 'aminer' # optional default: aminer

LogResourceList:
        - 'file:///home/ubuntu/git/test.txt'

Parser:
       - id: a_elem
         type: VariableByteDataModelElement
         name: 'a'
         args: 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ 0123456789_-.[]"\\'

       - id: b_elem
         type: FixedDataModelElement
         name: 'b'
         args: 'null'

       - id: z_elem
         type: VariableByteDataModelElement
         name: 'z'
         args: 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ 0123456789_-.[]'

       - id: y_elem
         type: VariableByteDataModelElement
         name: 'y'
         args: 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ 0123456789_-.[]'

       - id: json
         start: True
         type: JsonModelElement
         name: 'model'
         key_parser_dict:
            a: a_elem
            x:
              z: z_elem
              y: y_elem
              b: b_elem

Input:
        verbose: True # optional
        multi_source: False # optional
        timestamp_paths: "/model/accesslog/time"
        json_format: True

Analysis:

EventHandlers:
        - id: "stpe"
          json: true # optional default: false
          type: "StreamPrinterEventHandler"
```

In the data, the first atom should match the parser, the second atom fails because there is a ":" in the string, which is not part of the VariableBytDataModelElement allowed characters.

```
{"a": "works", "x": {"z": "test", "y": "test", "b": null}}
{"a": "fails", "x": {"z": "test", "y": "string fails because of colon: end of string", "b": null}}
```

It is just an idea, but it could have to do with taking the length of encoded vs unencoded strings, which is done in the json parsing element. For example, here the lengths are different (this could also be unrelated to this issue, but relevant for parsing logs with special characters):

```
>>> a = "special character: ö"
>>> len(a)
20
>>> len(a.encode())
21
```

**Comment by dITySoftware** [comment]
This issue is known to me and it happens, because internally the JsonModelElement works with the parsed json data and can only estimate how much data has to be updated from the MatchContext. For example multiple spaces or newlines are removed when parsing to json.

Currently i do not have a solution for this problem and it only affects the debugging.

**Comment by dITySoftware** [comment]
fixed in #606.
New output:
2021-04-04 09:47:00 Unparsed atom received
VerboseUnparsedAtomHandler: "None" (1 lines)
  {
  "DebugLog": [
    "Starting match update on b'{\"a\": \"fails\", \"x\": {\"z\": \"test\", \"y\": \"string fails because of colon: end of string\", \"b\": null}}'",
    "Removed b'{\"a\": \"fails', remaining 86 bytes",
    "Removed b'\", \"x\": {\"z\": \"test', remaining 67 bytes",
    "Removed b'\", \"y\": \"string fails because of colon: end of string\"', remaining 13 bytes",
    "Removed b', \"b\": null', remaining 2 bytes",
    "Shortest unmatched data was b'}}'",
    ""
  ],
  "LogData": {
    "RawLogData": [
      "{\"a\": \"fails\", \"x\": {\"z\": \"test\", \"y\": \"string fails because of colon: end of string\", \"b\": null}}"
    ],
    "Timestamps": [
      1617522420.15
    ],
    "DetectionTimestamp": 1617522420.15,
    "LogLinesCount": 1
  },
  "AnalysisComponent": {
    "AnalysisComponentIdentifier": null,
    "AnalysisComponentType": "VerboseUnparsedAtomHandler",
    "AnalysisComponentName": null,
    "Message": "Unparsed atom received",
    "PersistenceFileName": null
  }
}

**Comment by landauermax** [comment]
Yes, I think this is a good improvement.
